prog/dftb+/lib_dftb/rangeseparated.F90

```diff
@@ -693,16 +699,181 @@ subroutine updateHamiltonianBlock(descM, descN, pSma, pSbN, pPab)
       real(dp), dimension(:,:), pointer :: pHmn
 
       pHmn => tmpHH(descM(iStart):descM(iEnd), descN(iStart):descN(iEnd))
-      if(this%tSpin) then
-        pHmn(:,:) = pHmn - 0.25_dp * gammaTot * matmul(matmul(pSma, pPab), pSbn)
-      else
-        pHmn(:,:) = pHmn - 0.125_dp * gammaTot * matmul(matmul(pSma, pPab), pSbn)
-      end if
+      pHmn(:,:) = pHmn - 0.125_dp * this%nSpin * gammaTot * matmul(matmul(pSma, pPab), pSbn)
 
     end subroutine updateHamiltonianBlock
 
   end subroutine addLrHamiltonianNeighbour
 
+  subroutine addLrHamiltonianMatrix(this, iSquare, orb, overlap, densSqr, HH)
+
+    !> class instance
+    type(RangeSepFunc), intent(inout) :: this
+
+    !> Position of each atom in the rows/columns of the square matrices. Shape: (nAtom)
+    integer, dimension(:), intent(in) :: iSquare
+
+    !> Orbital information.
+    type(TOrbitals), intent(in) :: orb
+
+    real(dp), intent(in) :: overlap(:,:)
+    real(dp), intent(in) :: densSqr(:,:)
+
+    real(dp), intent(inout) :: HH(:,:)
+
+    real(dp), allocatable :: Smat(:,:)
+    real(dp), allocatable :: Dmat(:,:)
+    real(dp), allocatable :: LRgammaAO(:,:)
+    real(dp), allocatable :: Hlr(:,:)
+
+    integer :: nAOs
+
+    nAOs = size(overlap,dim=1)
+
+    allocate(Smat(nAOs,nAOs))
+    allocate(Dmat(nAOs,nAOs))
+    allocate(LRgammaAO(nAOs,nAOs))
+    allocate(Hlr(nAOs,nAOs))
+
+    call allocateAndInit(this, iSquare, orb, overlap, densSqr, HH, Smat, Dmat, LRgammaAO)
+    call evaluateHamiltonian(this, Smat, Dmat, LRgammaAO, Hlr)
+    HH(:,:) = HH(:,:) + Hlr(:,:)
```

Comment:
Remove `(:,:)` from the right hand side! (Also on several places further down...)